src/engine/guidance/post_processing.cpp

```diff
@@ -982,10 +992,14 @@ void trimShortSegments(std::vector<RouteStep> &steps, LegGeometry &geometry)
     // all zero-length instructions
     if (next_to_last_step.distance <= 1 && steps.size() > 2)
     {
-        geometry.locations.pop_back();
-        geometry.annotations.pop_back();
-        geometry.osm_node_ids.pop_back();
         geometry.segment_offsets.pop_back();
+        while (geometry.locations.size() > (geometry.segment_offsets.back() + 1))
+        {
+            geometry.locations.pop_back();
+            geometry.annotations.pop_back();
```

Comment:
Here you can use `geometry.locations.resize(geometry.segment_offsets.back() + 1)` which won't discard the first n elements.